core/src/main/scala/chisel3/properties/Property.scala

```diff
@@ -40,38 +67,72 @@ private[chisel3] trait PropertyType[T] {
+    makeSimple[BigInt](_ => fir.IntegerPropertyType, fir.IntegerPropertyLiteral(_))
 
   @implicitAmbiguous("unable to infer Property type. Please specify it explicitly in square brackets on the LHS.")
-  implicit val classPropertyTypeInstance = new PropertyType[ClassType] {
-    override def getPropertyType(value: Option[ClassType]): ir.PropertyType = ir.ClassPropertyType(value.get.name)
-  }
+  implicit val classPropertyTypeInstance = makeSimple[ClassType](
+    value => fir.ClassPropertyType(value.get.name),
+    value => fir.StringPropertyLiteral(value.name)
+  )
 
-  implicit val stringPropertyTypeInstance = new PropertyType[String] {
-    override def getPropertyType(value: Option[String]): ir.PropertyType = ir.StringPropertyType
-  }
+  implicit val stringPropertyTypeInstance =
+    makeSimple[String](_ => fir.StringPropertyType, fir.StringPropertyLiteral(_))
+
+  implicit val boolPropertyTypeInstance =
+    makeSimple[Boolean](_ => fir.BooleanPropertyType, fir.BooleanPropertyLiteral(_))
 
-  implicit val boolPropertyTypeInstance = new PropertyType[Boolean] {
-    override def getPropertyType(value: Option[Boolean]): ir.PropertyType = ir.BooleanPropertyType
+  implicit def propertyTypeInstance[T](implicit pte: PropertyType[T]) = new PropertyType[Property[T]] {
+    type Type = T
+    override def getPropertyType(value: Option[Property[T]]): fir.PropertyType = pte.getPropertyType(None)
+    override def convert(value:         Underlying, ctx: ir.Component, info: SourceInfo): fir.Expression =
+      ir.Converter.convert(value, ctx, info)
+    type Underlying = ir.Arg
+    override def convertUnderlying(value: Property[T]) = value.ref
   }
 
-  implicit def sequencePropertyTypeInstance[A: PropertyType, F[_] <: Seq[_]] = new PropertyType[F[A]] {
-    override def getPropertyType(value: Option[F[A]]): ir.PropertyType =
-      ir.SequencePropertyType(implicitly[PropertyType[A]].getPropertyType(None))
+  implicit def sequencePropertyTypeInstance[A, F[A] <: Seq[A]](implicit tpe: PropertyType[A]) = new PropertyType[F[A]] {
+    type Type = F[tpe.Type]
+    override def getPropertyType(value: Option[F[A]]): fir.PropertyType =
+      fir.SequencePropertyType(implicitly[PropertyType[A]].getPropertyType(None))
+
+    type Underlying = Seq[tpe.Underlying]
+    override def convert(value: Underlying, ctx: ir.Component, info: SourceInfo): fir.Expression =
+      fir.SequencePropertyValue(tpe.getPropertyType(None), value.map(tpe.convert(_, ctx, info)))
+    override def convertUnderlying(value: F[A]) =
+      value.map(tpe.convertUnderlying(_))
   }
+
+  implicit def mapPropertyTypeInstance[A, F[A] <: Map[String, A]](implicit tpe: PropertyType[A]) =
+    new PropertyType[F[A]] {
+      type Type = F[tpe.Type]
+      override def getPropertyType(value: Option[F[A]]): fir.PropertyType =
+        fir.MapPropertyType(implicitly[PropertyType[A]].getPropertyType(None))
+      type Underlying = Map[String, tpe.Underlying]
```

Comment:
```suggestion
      type Underlying = VectorMap[String, tpe.Underlying]
```
We definitely should use an ordered map here.